uc-crux-llvm/src/UCCrux/LLVM/Classify.hs

```diff
@@ -0,0 +1,643 @@
+  | ArgFreeUnallocated
+  | ArgFreeBadOffset
+  | ArgWriteUnmapped
+  | ArgReadUninitialized
+  | ArgReadUninitializedOffset
+  | ArgPointerConstOffset
+  deriving (Eq, Ord)
+
+diagnose :: MissingPreconditionTag -> Text
+diagnose =
+  \case
+    ArgWriteBadAlignment -> "Write to a pointer with insufficient alignment in argument"
+    ArgReadBadAlignment -> "Read from a pointer with insufficient alignment in argument"
+    ArgFreeUnallocated -> "`free` called on an unallocated pointer in argument"
+    ArgFreeBadOffset -> "`free` called on pointer with nonzero offset in argument"
+    ArgWriteUnmapped -> "Write to an unmapped pointer in argument"
+    ArgReadUninitialized -> "Read from an uninitialized pointer in argument"
+    ArgReadUninitializedOffset -> "Read from an uninitialized pointer calculated from a pointer in argument"
+    ArgPointerConstOffset -> "Addition of a constant offset to a pointer in argument"
+
+prescribe :: MissingPreconditionTag -> Text
+prescribe =
+  ("Prescription: Add a precondition that " <>)
+    . \case
+      ArgWriteBadAlignment -> "the pointer is sufficiently aligned."
+      ArgReadBadAlignment -> "the pointer is sufficiently aligned."
+      ArgFreeUnallocated -> "the pointer points to initialized memory."
+      ArgFreeBadOffset -> "the pointer points to initialized memory."
+      ArgWriteUnmapped -> "the pointer points to allocated memory."
+      ArgReadUninitialized -> "the pointer points to initialized memory."
+      ArgReadUninitializedOffset -> "the pointer points to initialized memory."
+      ArgPointerConstOffset -> "the allocation is at least that size."
+
+-- | There was an error and we know roughly what sort it was, but we still can't
+-- do anything about it.
+data Unfixable
+  = AddSymbolicOffsetToArgPointer
+  deriving (Eq, Ord, Show)
+
+ppUnfixable :: Unfixable -> Text
+ppUnfixable =
+  \case
+    AddSymbolicOffsetToArgPointer -> "Addition of a symbolic offset to pointer in argument"
+
+-- | We know how we *could* fix this in theory, but haven't implemented it yet.
+data Unfixed
+  = UnfixedArgPtrOffsetArg
+  deriving (Eq, Ord, Show)
+
+ppUnfixed :: Unfixed -> Text
+ppUnfixed =
+  \case
+    UnfixedArgPtrOffsetArg -> "Addition of an offset from argument to a pointer in argument"
+
+-- | Possible sources of uncertainty, these might be true or false positives
+data Uncertainty
+  = -- | We don't (yet) know what to do about this error
+    UUnclassified (Doc Void)
+  | UUnfixable Unfixable
+  | UUnfixed Unfixed
+  | -- | Simulation, input generation, or classification encountered
+    -- unimplemented functionality
+    UUnimplemented (Panic Unimplemented)
+  | -- | This @Pred@ was not annotated
+    UMissingAnnotation Crucible.SimError
+  | -- | A user assertion failed, but symbolically
+    UFailedAssert !What4.ProgramLoc
+  deriving (Show)
+
+partitionUncertainty ::
+  [Uncertainty] -> ([Crucible.SimError], [What4.ProgramLoc], [Panic Unimplemented], [Doc Void], [Unfixed], [Unfixable])
```

Comment:
I know how this is used, so I'm not hugely concerned, but turning a big tuple like that into a datatype might be more comprehensible if we ever need to extend it

Reply:
https://github.com/GaloisInc/crucible/issues/691